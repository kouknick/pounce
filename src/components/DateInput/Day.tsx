import React from 'react';
import Box from '../Box';
import AbstractButton from '../AbstractButton';
import Cell from './Cell';
import dayjs, { Dayjs } from 'dayjs';
import { noop } from '../../utils/helpers';

export interface DayProps {
  day?: number;
  isLastRow?: boolean;
  month: number;
  year: number;
  daySelected?: Dayjs;
  dayRangeSelected?: [Dayjs?, Dayjs?];
  onDaySelect?: (date: Dayjs) => void;
}

const Day: React.FC<DayProps> = ({
  day,
  month,
  year,
  daySelected,
  isLastRow,
  dayRangeSelected,
  onDaySelect = noop,
}) => {
  if (!day) {
    return isLastRow ? null : (
      <Cell
        display="table-cell"
        verticalAlign="middle"
        textAlign="center"
        cursor="pointer"
        position="relative"
        aria-disabled="true"
      />
    );
  }
  const date = React.useMemo(() => dayjs().month(month).date(day).year(year), [year, day, month]);
  const onDaySelectClick = React.useCallback(
    e => {
      e.preventDefault();
      onDaySelect(date);
    },
    [date, onDaySelect]
  );

  const isSelected = React.useMemo(() => {
    if (!daySelected && !dayRangeSelected) {
      return false;
    }
    if (!date) {
      return false;
    }

    if (daySelected) {
      return date.isSame(daySelected, 'date');
    }
    // @ts-ignore
    const [start, end] = dayRangeSelected;
    return (start && date.isSame(start, 'date')) || (end && date.isSame(end, 'date'));
  }, [date, daySelected, dayRangeSelected]);

  const isWithinRange = React.useCallback(() => {
    if (!dayRangeSelected) {
      return false;
    }
    // @ts-ignore
    const [start, end] = dayRangeSelected;
    return start && date.isAfter(start, 'date') && end && date.isBefore(end, 'date');
  }, [date, dayRangeSelected]);

  return (
    <Cell
      display="table-cell"
      verticalAlign="middle"
      textAlign="center"
      cursor="pointer"
      position="relative"
      aria-label={date.format('dd MMM DD YYYY')}
      role="gridcell"
      aria-busy={isWithinRange()}
      aria-selected={isSelected}
    >
      <Box aria-placeholder="true" tabIndex={-1} pointerEvents="none" />
      <AbstractButton
        padding={2}
        borderRadius="circle"
        position="relative"
        zIndex={2}
        onClick={onDaySelectClick}
        // @ts-ignore
        aria-relevant
        width="37px"
        height="37px"
      >
        {day}
      </AbstractButton>
    </Cell>
  );
};

export default React.memo(Day);
