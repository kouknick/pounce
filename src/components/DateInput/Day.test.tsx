import React from 'react';
import { renderWithTheme, fireEvent } from 'test-utils';
import dayjs from 'dayjs';
import Day from './Day';

const month = 10;
const year = 2020;
const day = 1;
const date = dayjs().date(day).month(month).year(year);
const endDate = date.add(1, 'month');

describe('Day', () => {
  it('renders a disabled cell', async () => {
    const { container } = await renderWithTheme(<Day month={month} year={year} />);
    expect(container).toMatchSnapshot();
  });

  it('renders a cell when date is present', async () => {
    const { container } = await renderWithTheme(<Day day={day} month={month} year={year} />);
    expect(container).toMatchSnapshot();
  });

  it('renders a cell selected', async () => {
    const { container } = await renderWithTheme(
      <Day daysSelected={[date]} day={day} month={month} year={year} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders a selected date', async () => {
    const { container } = await renderWithTheme(
      <Day daysSelected={[date, endDate]} day={day} month={month} year={year} />
    );
    expect(container).toMatchSnapshot();
  });

  it('allows selecting a date', async () => {
    const mock = jest.fn();
    const { findByLabelText } = await renderWithTheme(
      <Day day={day} month={month} year={year} onDaySelect={mock} />
    );
    const cell = await findByLabelText('Su Nov 01 2020');
    const btn = cell.querySelector('button');
    await fireEvent.click(btn);
    expect(mock).toHaveBeenCalled();
    const date = mock.mock.calls[0][0];
    expect(date.format('DD/MM/YYYY')).toBe('01/11/2020');
  });
});
