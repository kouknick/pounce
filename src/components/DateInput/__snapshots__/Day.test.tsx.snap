// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day renders a cell selected 1`] = `
.pounce-0 {
  display: table-cell;
  padding: 8px;
  text-align: center;
  border-radius: 99999px;
  cursor: pointer;
  vertical-align: middle;
}

.pounce-0[aria-selected=true],
[data-selected]>.pounce-0 {
  background-color: #0081c5;
}

.pounce-0:hover {
  background-color: #11a5f3;
}

<div>
  <div
    aria-disabled="false"
    aria-label="Su Nov 01 2020"
    aria-selected="true"
    class="pounce-0"
    role="gridcell"
  >
    1
  </div>
</div>
`;

exports[`Day renders a cell when date is present 1`] = `
.pounce-0 {
  display: table-cell;
  padding: 8px;
  text-align: center;
  border-radius: 99999px;
  cursor: pointer;
  vertical-align: middle;
}

.pounce-0[aria-selected=true],
[data-selected]>.pounce-0 {
  background-color: #0081c5;
}

.pounce-0:hover {
  background-color: #11a5f3;
}

<div>
  <div
    aria-disabled="false"
    aria-label="Su Nov 01 2020"
    aria-selected="false"
    class="pounce-0"
    role="gridcell"
  >
    1
  </div>
</div>
`;

exports[`Day renders a disabled cell 1`] = `
.pounce-0 {
  display: table-cell;
  padding: 8px;
  text-align: center;
  border-radius: 99999px;
  cursor: pointer;
  vertical-align: middle;
}

.pounce-0[aria-selected=true],
[data-selected]>.pounce-0 {
  background-color: #0081c5;
}

.pounce-0:hover {
  background-color: #11a5f3;
}

<div>
  <div
    aria-disabled="true"
    class="pounce-0"
  />
</div>
`;
